kernel/posix/pthread.c

```diff
@@ -0,0 +1,550 @@
+		return -1;
+	}
+
+	*oldstate = pthread->cancel_state;
+	pthread_mutex_lock(&pthread->cancel_lock);
+	pthread->cancel_state = state;
+
+	if (state == PTHREAD_CANCEL_DISABLE) {
+		_thread_essential_set();
+	} else {
+		_thread_essential_clear();
+	}
+
+	pthread_mutex_unlock(&pthread->cancel_lock);
+
+	if (state == PTHREAD_CANCEL_ENABLE && pthread->cancel_pending) {
+		pthread_exit((void *)PTHREAD_CANCELED);
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Cancel execution of a thread.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_cancel(pthread_t pthread)
+{
+	struct posix_thread *thread = (struct posix_thread *) pthread;
+
+	if (thread == NULL || thread->state == PTHREAD_TERMINATED) {
+		errno = ESRCH;
+		return -1;
+	}
+
+	pthread_mutex_lock(&thread->cancel_lock);
+	thread->cancel_pending = 1;
+
+	if (thread->cancel_state == PTHREAD_CANCEL_ENABLE) {
+		if (thread->state == PTHREAD_DETACHED) {
+			thread->state = PTHREAD_TERMINATED;
+			pthread_mutex_unlock(&thread->cancel_lock);
+			k_thread_abort((k_tid_t) thread);
+		} else {
+			thread->retval = PTHREAD_CANCELED;
+			pthread_mutex_unlock(&thread->cancel_lock);
+		}
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Set thread scheduling policy and parameters.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_setschedparam(pthread_t pthread, int policy,
+			  const struct sched_param *param)
+{
+	k_tid_t *thread = (k_tid_t *)pthread;
+	int new_prio, cur_prio;
+
+	if (thread == NULL) {
+		errno = ESRCH;
+		return -1;
+	}
+
+	if (policy != SCHED_RR && policy != SCHED_FIFO) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	new_prio = posix_to_zephyr_priority(param->priority, policy);
+	cur_prio = k_thread_priority_get(*thread);
+
+	if (is_posix_prio_valid(new_prio, policy) == false ||
+	    new_prio == cur_prio) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	k_thread_priority_set(*thread, new_prio);
+
+	return 0;
+}
+
+/**
+ * @brief Initialise threads attribute object
+ *
+ * See IEEE 1003.1
+ */
+int pthread_attr_init(pthread_attr_t *attr)
+{
+
+	if (attr->initialized == true) {
+		errno = EBUSY;
+		return -1;
+	}
+
+	*attr = init_pthread_attrs;
+
+	return 0;
+}
+
+/**
+ * @brief Get thread scheduling policy and parameters
+ *
+ * See IEEE 1003.1
+ */
+int pthread_getschedparam(pthread_t pthread, int *policy,
+			  struct sched_param *param)
+{
+	k_tid_t thread = (k_tid_t)pthread;
+	u32_t  priority = k_thread_priority_get(thread);
+
+	param->priority = zephyr_to_posix_priority(priority, policy);
+	return 0;
+}
+
+/**
+ * @brief Terminate calling thread.
+ *
+ * See IEEE 1003.1
+ */
+void pthread_exit(void *retval)
+{
+	struct posix_thread *self = (struct posix_thread *)pthread_self();
+
+	pthread_mutex_lock(&self->state_lock);
+
+	if (self->state == PTHREAD_JOINABLE) {
+		self->retval = retval;
+		self->state = PTHREAD_EXITED;
+		self->retval = retval;
+		pthread_cond_broadcast(&self->state_cond);
+		pthread_mutex_unlock(&self->state_lock);
+	} else {
+		pthread_mutex_unlock(&self->state_lock);
+		self->state = PTHREAD_TERMINATED;
```

Comment:
Shouldn't the state be set before the mutex is unlocked?

Reply:
Thanks for catching this.

Yes we should change the state only when it is protected through state_lock.
I will make this change.